### DISCUSSION TITLE
switch between scripts with buttons for FLL?

### DISCUSSION THREAD

**Discussion by jediforri** [pybricks/support#1307] [General]
I've written some scripts for different FLL missions.
How can I use the buttons on my spike hub to pick the script I want to run?
And then can I start and stop scripts using the circle button?
Thanks

**Comment by dlech** [comment]
Did you try the `hub_menu()` function?

https://docs.pybricks.com/en/stable/tools/index.html#pybricks.tools.hub_menu

**Reply by jediforri** [reply]
Oh that looks easy, thanks.

Is there a way to cancel a scrip that's running by pressing a button if we need to grab the robot because it gets stuck? (this hopefully will not happen)

**Reply by dlech** [reply]
Pressing the center button will raise a `SystemExit` exception which will stop everything (as long as the exception is not caught).

**Reply by rajrao** [reply]
Our FLL team wrote some code for the menu (when hub_menu was not available). But they have found it to be more flexible. Example: Bluetooth button will stop the entire code, center button will stop the currently running program.
It was a good learning experience too for the kids.

https://github.com/coder-ella/lego/blob/main/pybricks/masterpiece/A_main_menu.py

**Reply by laurensvalk** [reply]
Thanks Raj!

**Reply by jediforri** [reply]
can we use hub_menu with multi_task? We cannot figure it out without getting an error that we can't use hub_menu once we multitask?

We want to use hub_menu to launch different multi_tasks

**Reply by laurensvalk** [reply]
The `hub_menu` currently isn't set up for multitasking. Would this be useful for a future release?

Note that the hub menu is not a special function. It is just a convenience function that mixes buttons, lights and waiting. You could fairly easily make something yourself that will work with multitasking.

**Reply by rajrao** [reply]
@jediforri you can do the following

```
while True:
        selected = hub_menu("1","2","x")
        if selected == "1":
            run_task(slot1())
        elif selected == "2":
            run_task(slot2())
        else:
             break
```
Slot1 and slot2 are async methods. That looks like this:
```
async def slot1():
    await multitask(motor_attach_left.run_angle(500, -360)
    , motor_attach_right.run_angle(500, -360))
    
    await multitask(motor_attach_left.run_angle(500, 360),
    motor_attach_right.run_angle(500, 360))
```

**Reply by jediforri** [reply]
yes, this would be a great feature to have in future releases, thanks

**Reply by jediforri** [reply]
> The `hub_menu` currently isn't set up for multitasking. Would this be useful for a future release?
> 
> Note that the hub menu is not a special function. It is just a convenience function that mixes buttons, lights and waiting. You could fairly easily make something yourself that will work with multitasking.

is multitasking supported with the hub_menu now? Hope so! If not, is that something I could we could build using pybricks?

**Reply by laurensvalk** [reply]
Just to clarify, you _can_ use the `hub_menu` technique to select programs that _use_ multitasking. So in [this tutorial](https://pybricks.com/project/spike-hub-menu/) it would be fine if, say, the `hello_world` program used multitasking.

It's just that the `hub_menu` itself doesn't run in parallel to something else. But I don't think this is what you're after? Usually the robot isn't driving while you're selecting something on the display.

Otherwise, perhaps you can share an example of what you'd like to do?

**Reply by jediforri** [reply]
[rajrao](https://github.com/rajrao)'s code helped me with what I wanted to do - thank you!

**Comment by laurensvalk** [comment]
We've just added a tutorial that also answers this question. It works both with Python and block-based coding.

[![spike-hubmenu-yt](https://github.com/pybricks/support/assets/12326241/eeba4d24-4d68-40fd-8718-71e11e6634cf)](https://pybricks.com/project/spike-hub-menu/)

**Comment by jediforri** [comment]
Ok, trying again. Help appreciated with this problem.

On beta.pybricks, this does not throw an error, but does not move the wheels either...

```
async def mission1():
    await db.straight( 10 )

async def mission2():
    await db.straight( -10 )

# given abcd, and b, return bcda
def move_to_front(my_list, value):
    index = my_list.index(value)
    shifted_list = my_list[index:] + my_list[:index]
    return shifted_list

missions = ["1", "2"]
while True:
    selected = hub_menu( *missions )

    if selected == "1":
        mission1()
        missions = move_to_front(missions, "2")
    elif selected == "2":
        mission2()
        missions = move_to_front(missions, "1")
```

**Reply by rajrao** [reply]
Does mission1() need to be wrapped in a run task?

**Reply by jediforri** [reply]
Fixed it!! Thank you.

```

async def mission1():
    await db.straight( 1000 )

async def mission2():
    await db.straight( -1000 )

# given abcd, and b, return bcda
def move_to_front(my_list, value):
    index = my_list.index(value)
    shifted_list = my_list[index:] + my_list[:index]
    return shifted_list

missions = ["1", "2"]
while True:
    selected = hub_menu( *missions )

    if selected == "1":
        run_task(mission1())
        missions = move_to_front(missions, "2")
    elif selected == "2":
        run_task(mission2())
        missions = move_to_front(missions, "1")
```
